### PR TITLE
feat: Add mobile endpoint for Web Atoms

### PR DESCRIPTION
**Pull request by dpgraham**
Implements an endpoint `mobile: webAtoms` which calls the Espresso web atoms API. The call provides JSON that provides an optional webView element, and then list of atoms to execute.

e.g.)

```javascript
{
      webviewElement: webviewEl.value,
      forceJavascriptEnabled: true,
      methodChain: [
        {name: 'withElement', atom: {name: 'findElement', locator: {using: 'ID', value: 'i_am_a_textbox'}}},
        {name: 'perform', atom: {name: 'webKeys', args: 'Hello world'}},
        {name: 'withElement', atom: {name: 'findElement', locator: {using: 'ID', value: 'i am a link'}}},
        {name: 'perform', atom: 'webClick'},
      ]
    }
```

calls...

```java
onWebView(theWebView)
  .forceJavascriptEnabled(true)
  .withElement(findElement(Locator.ID, "i_am_a_textbox"))
  .perform(webKeys("Hello world))
  .withElement(findElement(Locator.ID, "i am a link"))
  .perform(webClick())
```

Also added KReflectionUtils which implements reflection using Kotlin.

### CODE REVIEW

**Comment by jlipps** [comment]
why expose this as a mobile command? i thought the idea was that we would switch context into the webview, and from then on all commands would ultimately generate atoms internally, but the user wouldn't know that as it's just an implementation detail?

**Comment by dpgraham** [comment]
@jlipps this is a stopgap. I wanted to add this capability to see if there's interest in it. Also because the other option would require a lot more effort and I would like this available in the meantime.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import io.appium.espressoserver.lib.helpers.KReflectionUtils.invokeInstanceMethod
+import io.appium.espressoserver.lib.helpers.KReflectionUtils.invokeMethod
+import io.appium.espressoserver.lib.model.Element
+import io.appium.espressoserver.lib.viewmatcher.WithView.withView
+
+class WebAtoms : RequestHandler<WebAtomsParams, Void> {
+
+    @Throws(AppiumException::class)
+    override fun handle(webAtomsParams: WebAtomsParams): Void? {
+        var webViewInteraction: WebInteraction<*>
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt]: how about

```
var webViewInteraction = onWebView();

...

webAtomsParams.webviewElement?.let{
  webViewInteraction = webViewInteraction(withView(Element.getViewById(it)))
}
```

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        // Iterate through methodsChain and call the atoms
+        for (method in webAtomsParams.methodChain) {
+            logger.info("Calling method '${method.name}' with atom '${method.atom}'");
+
+            logger.info("Calling atom '${method.atom.name!!}' with args '${method.atom.args.joinToString(", ")}'")
+            val atom = invokeMethod(DriverAtoms::class, method.atom.name!!, *method.atom.args.toTypedArray());
+
+            logger.info("Calling interaction '${method.name}' with the atom")
+            var res:Any?
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt]: same comment as above

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            if (atom != null) {
+                res = invokeInstanceMethod(webViewInteraction, method.name, atom);
+            } else {
+                res = invokeInstanceMethod(webViewInteraction, method.name);
+            }
+
+            if (res is WebInteraction<*>) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt]: is res in not ... then throw. else is redundant

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/KReflectionUtils.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import java.lang.IllegalArgumentException
+import java.lang.reflect.InvocationTargetException
+import kotlin.reflect.KClass
+import kotlin.reflect.KFunction
+import kotlin.reflect.full.createType
+import kotlin.reflect.full.functions
+import kotlin.reflect.full.memberFunctions
+import kotlin.reflect.jvm.javaType
+
+object KReflectionUtils {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/KReflectionUtils.kt]: what is `K`? can we have a better class name?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/KReflectionUtils.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            if (func.name == methodName) {
+
+                // If the length of func parameters provided isn't same as expected, go to next.
+                val funcParams = func.parameters
+                if (funcParams.size != providedParams.size) {
+                    continue
+                }
+
+                // Look through function parameters and do an enum hack to translate strings to enums
+                for (index in 0 until funcParams.size) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/KReflectionUtils.kt]: how about using https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/for-each-indexed.html ?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/KReflectionUtils.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                    // Hack Enum Case
+                    // If function param is Enum and provided param is String, try `enumValueOf` on that String value
+                    try {
+                        val jFuncType = funcParamType.javaType as Class<*>
+                        if (jFuncType.isEnum && providedParam is String) {
+                            val enumValueOf = jFuncType.getDeclaredMethod("valueOf", String::class.java)
+                            treatedParams.set(index, enumValueOf(null, providedParam.toUpperCase()))
+                            continue;
+                        }
+                    } catch (e:Exception) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/KReflectionUtils.kt]: do we have more precise exception class?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/KReflectionUtils.kt (hunk 4 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/KReflectionUtils.kt]: Kotlin. It follows the convention they use to distinguish Kotlin from Java (KClass, KType, etc...)

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtom.kt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                val args = jsonObj.get("args")
+                if (args != null) {
+                    if (args.isJsonPrimitive) {
+                        webAtom.args = Collections.singletonList(args.asString)
+                    } else if (args.isJsonArray){
+                        val argsAsList = ArrayList<Any>()
+                        for (arg in args.asJsonArray) {
+                            if (arg.isJsonPrimitive) {
+                                val argPrimitive = arg.asJsonPrimitive
+                                if (argPrimitive.isBoolean) argsAsList.add(argPrimitive.asBoolean)
```
[the remark this answers is not in the document]
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtom.kt]: Yeah. The syntax is different, but in this case I couldn't find anything for the `JsonPrimitive` that returns the type. I had to go check the types individually.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtomsParams.kt [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package io.appium.espressoserver.lib.model.web
+
+import io.appium.espressoserver.lib.model.AppiumParams
+import java.util.*
```
[the remark this answers is not in the document]
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtomsParams.kt]: I took it out altogether and switched to using Kotlin's [emptyList()](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/empty-list.html)

[the remark this answers is not in the document]
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtomsParams.kt]: I think we need to enable Kotlin on Codacy... I'm not seeing any linting of .kt files

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/KReflectionUtils.kt (hunk 6 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/KReflectionUtils.kt]: Yeah, I'm going to change it to the broader `ReflectiveTypeException` class. That should cover it.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtom.kt [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                                throw JsonParseException("'${arg}' is not a valid 'arg' type");
+                            }
+                        }
+                        webAtom.args = argsAsList
+                    }
+                }
+
+                // Parse the Locator shorthand
+                val locator = jsonObj.get("locator")
+                if (locator != null) {
```
[the remark this answers is not in the document]
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtom.kt]: I'm just learning about `let` know. Seems like the best choice for null checking.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt (hunk 3 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt]: Nvm: Your way works too. Smart-casting covers it.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtom.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtom.kt]: Kotlin, actually has `when` operator, which might be in use here. Check https://superkotlin.com/kotlin-when-statement/

File: lib/commands/general.js [hunk 10]
```diff
@@ -95,6 +95,10 @@ commands.mobileNavigateTo = async function (opts = {}) {
   });
 };
 
+commands.mobileWebAtoms = async function (opts = {}) {
```
[the remark this answers is not in the document]
Comment (dpgraham) [reply lib/commands/general.js]: I'll add it to the mobile documents section in appium.io

[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply lib/commands/general.js]: do we have any required fields in opts?

File: lib/commands/general.js [hunk 11]
```diff
@@ -95,6 +95,25 @@ commands.mobileNavigateTo = async function (opts = {}) {
...
+/**
+ * Runs a chain of Espresso web atoms
+ *
+ * Takes JSON of the form
+ *
+ * {
+ *   "webviewEl": "<ELEMENT_ID>", // optional webview element to operate on
+ *   "forceJavascriptEnabled": true|false, // if webview disables javascript, webatoms won't work, this forces it
+ *   "methods": [
+ *     {"name": "methodName", "atom": {"name": "atomName", "args": ["arg1", "arg2", ...]}},
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: Would you mind to also add a link containing a list of available atoms?

Comment (dpgraham) [reply lib/commands/general.js]: Done

File: lib/commands/general.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (dpgraham) [reply lib/commands/general.js]: Just `methodChain`.... I'll add that.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        if (webAtomsParams.forceJavascriptEnabled) {
+            webViewInteraction.forceJavascriptEnabled()
+        }
+
+        // Iterate through methodsChain and call the atoms
+        for (method in webAtomsParams.methodChain) {
+            val atom = invokeMethod(DriverAtoms::class, method.atom.name, *method.atom.args.toTypedArray());
+
+            logger.info("Calling interaction '${method.name}' with the atom '${method.atom}'")
+            val args = if (atom == null) emptyList<Any>() else atom;
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt]: there is also `?:` operator ;)

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            val res = invokeInstanceMethod(webViewInteraction, method.name, args);
+
+            if (!(res is WebInteraction<*>)) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/WebAtoms.kt]: `!is` is also valid

**Pull request merged**